File: Problems/0600_0699/0606_Construct_String_from_Binary_Tree/Project_C/lib/operate_treenode.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <time.h>

#include "../include/mylib.h"
#include "../include/treenode.h"
#include "../include/operate_treenode.h"

struct TreeNode *createTreeNode(char *flds)
{
    char *nums[65535];

    int nums_length = ml_split(flds, ",", nums, sizeof(nums)/sizeof(nums[0]));
    nums[nums_length] = NULL;

    return createSubTreeNode(nums, nums_length, 0, 0);
}

struct TreeNode *createSubTreeNode(char *flds[], int flds_length, int depth, int pos)
{
    if (flds_length == 0)
        return NULL;

    int cur_pos = 0;
    for (int i = 0; i < depth; ++i)
        cur_pos += (int)pow(2, i);
    
    if (cur_pos + pos > flds_length - 1)
        return NULL;
    
    if (strcmp(flds[cur_pos + pos], "null") == 0)
        return NULL;

    struct TreeNode *node = malloc(sizeof(struct TreeNode));
    node->val = strtol(flds[cur_pos + pos], NULL, 10);
    node->left = createSubTreeNode(flds, flds_length, depth + 1, 2*pos);
    node->right = createSubTreeNode(flds, flds_length, depth + 1, 2*pos + 1);

    return node;
}

#define MAX_STAIR_DEPTH     256
#define MAX_STRING_LENGTH   1024

char* tree_to_staircase_string(struct TreeNode *node) {
    char *resultList[MAX_STAIR_DEPTH];

    for (int i = 0; i < MAX_STAIR_DEPTH; ++i) {
        resultList[i] = malloc(sizeof(char)*MAX_STRING_LENGTH);
        resultList[i][0] = '\0';
    }

    tree_to_staircase_substring(node, resultList, 0);

    char* resultStr;
    int resultStr_length = 0;

    for (int i = 0; i < MAX_STAIR_DEPTH && resultList[i][0] != '\0'; ++i) {
        resultStr_length += strlen(resultList[i]) + 1;  // prefix '\n'
    }
    resultStr_length++;     // prefix '\0'

    resultStr = malloc(sizeof(char)*resultStr_length);
    resultStr[0] = '\0';

    for (int i = 0; i < MAX_STAIR_DEPTH && resultList[i][0] != '\0'; ++i) {
        strcat(resultStr, resultList[i]);
        strcat(resultStr, "\n");
    }
    resultStr[resultStr_length - 1] = '\0';

    return resultStr;
}

void tree_to_staircase_substring(struct TreeNode *node, char *resultList[], int n) {
    if (node == NULL)
        return;
    
    char val_to_str[MAX_STRING_LENGTH];

    if (resultList[n][0] == '\0') {
        sprintf(val_to_str, "(%d)", node->val);
    } else {
        sprintf(val_to_str, ",(%d)", node->val);
    }

    strcat(resultList[n], val_to_str);
    
    if (node->left != NULL) {
        tree_to_staircase_substring(node->left, resultList, n + 1);
    }
    if (node->right != NULL) {
        tree_to_staircase_substring(node->right, resultList, n + 1);
    }
}

char* tree2str(struct TreeNode *t) {
    if (t == NULL)
        return "";

    char *resultStr = NULL;

    if (t->left == NULL && t->right == NULL) {
        resultStr = calloc(10, sizeof(char));
        sprintf(resultStr, "%d", t->val);

    } else if (t->left != NULL && t->right == NULL) {
        char *resultStr_left = tree2str(t->left);
        resultStr = calloc(strlen(resultStr_left) + 10, sizeof(char));
        sprintf(resultStr, "%d(%s)", t->val, resultStr_left);

    } else if(t->left == NULL && t->right != NULL) {
        char *resultStr_right = tree2str(t->right);
        resultStr = calloc(strlen(resultStr_right) + 10, sizeof(char));
        sprintf(resultStr, "%d()(%s)", t->val, resultStr_right);

    } else if(t->left != NULL && t->right != NULL) {
        char *resultStr_left = tree2str(t->left);
        char *resultStr_right = tree2str(t->right);
        resultStr = calloc(strlen(resultStr_left) + strlen(resultStr_right) + 10, sizeof(char));
        sprintf(resultStr, "%d(%s)(%s)", t->val, resultStr_left, resultStr_right);

    }

    return resultStr;
}

File: Problems/0200_0299/0297_Serialize_and_Deserialize_Binary_Tree/Project_Go/solution/OperateTreeNode.go
package solution

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func createTreeNode(flds string) *TreeNode {
	return createSubTreeNode(strings.Split(flds, ","), 0, 0)
}

func createSubTreeNode(flds []string, depth int, pos int) *TreeNode {
	if len(flds) == 0 {
		return nil
	}

	curPos := 0
	for i := 0; i < depth; i++ {
		curPos += int(math.Pow(2.0, float64(i)))
	}

	if curPos+pos > len(flds)-1 {
		return nil
	}

	if flds[curPos+pos] == "null" {
		return nil
	}

	val, err := strconv.Atoi(flds[curPos+pos])
	if err != nil {
		fmt.Printf("createTreeNode() Error ... flds[%d] = %s", curPos+pos, flds[curPos+pos])
		return nil
	}

	/*
		node := new(TreeNode)
		node.Val = val
		node.Left = createSubTreeNode(flds, depth+1, 2*pos)
		node.Right = createSubTreeNode(flds, depth+1, 2*pos+1)
	*/
	node := &TreeNode{val, createSubTreeNode(flds, depth+1, 2*pos), createSubTreeNode(flds, depth+1, 2*pos+1)}
	return node
}

var resultList []string

func tree2staircaseString(node *TreeNode) string {
	resultList = []string{}
	outStr := setResultStr(node, 0)
	resultList = nil

	return outStr
}

func setResultStr(node *TreeNode, n int) string {
	if node == nil {
		return ""
	}

	if len(resultList) <= n {
		resultList = append(resultList, "("+strconv.Itoa(node.Val)+")")
	} else {
		resultList[n] = resultList[n] + ",(" + strconv.Itoa(node.Val) + ")"
	}

	if node.Left != nil {
		setResultStr(node.Left, n+1)
	}
	if node.Right != nil {
		setResultStr(node.Right, n+1)
	}

	resultStr := ""
	for _, s := range resultList {
		resultStr += s + "\n"
	}

	return resultStr
}

func tree2str(node *TreeNode) string {
	if node == nil {
		return ""
	}

	resultStr := strconv.Itoa(node.Val)

	if node.Left == nil && node.Right == nil {
		return resultStr
	}

	resultStr += "(" + tree2str(node.Left) + ")"
	if node.Right != nil {
		resultStr += "(" + tree2str(node.Right) + ")"
	}

	return resultStr
}
